src/tensorwaves/model/function.py

```diff
@@ -0,0 +1,117 @@
+"""Implementations of `.Function` and `.ParametrizedFunction`."""
+
+import inspect
+from typing import Callable, Dict, Iterable, Mapping, Tuple, Union
+
+import attr
+import numpy as np
+
+from tensorwaves.interface import (
+    DataSample,
+    Function,
+    Model,
+    ParameterValue,
+    ParametrizedFunction,
+)
+
+
+def _all_str(
+    _: "PositionalArgumentFunction", __: attr.Attribute, value: Iterable[str]
+) -> None:
+    if not all(map(lambda s: isinstance(s, str), value)):
+        raise TypeError(f"Not all arguments are of type {str.__name__}")
+
+
+def _all_unique(
+    _: "PositionalArgumentFunction", __: attr.Attribute, value: Iterable[str]
+) -> None:
+    argument_names = list(value)
+    if len(set(argument_names)) != len(argument_names):
+        duplicate_arguments = []
+        for arg_name in argument_names:
+            n_occurrences = argument_names.count(arg_name)
+            if n_occurrences > 1:
+                duplicate_arguments.append(arg_name)
+        raise ValueError(
+            f"There are duplicate argument names: {duplicate_arguments}"
+        )
+
+
+def _validate_arguments(
+    instance: "PositionalArgumentFunction", _: attr.Attribute, value: Callable
+) -> None:
+    if not callable(value):
+        raise TypeError("Function is not callable")
+    n_args = len(instance.argument_order)
+    signature = inspect.signature(value)
+    if len(signature.parameters) != n_args:
+        if len(signature.parameters) == 1:
+            parameter = next(iter(signature.parameters.values()))
+            if parameter.kind == parameter.VAR_POSITIONAL:
+                return
+        raise ValueError(
+            f"Lambdified function expects {len(signature.parameters)}"
+            f" arguments, but {n_args} sorted arguments were provided."
+        )
+
+
+def _to_tuple(argument_order: Iterable[str]) -> Tuple[str, ...]:
+    return tuple(argument_order)
+
+
+@attr.s(frozen=True)
+class PositionalArgumentFunction(Function):
+    """Wrapper around a function with positional arguments.
+
+    This class provides a :meth:`__call__` that can take a `.DataSample` for a
+    function with `positional arguments
+    <https://docs.python.org/3/glossary.html#term-positional-argument>`_. Its
+    :attr:`argument_order` redirect the keys in the `.DataSample` to the
+    argument positions in its underlying :attr:`function`.
+    """
+
+    function: Callable[..., np.ndarray] = attr.ib(
```

Comment:
I dont quite understand why these have to be exposed?

Reply:
I try to use [`attrs`](https://www.attrs.org) where possible, because it encourages writing composite classes that are easy to understand. But since `attrs` has a [specific take on 'private' attributes](https://www.attrs.org/en/stable/init.html#private-attributes), there isn't really any way to hide the attribute. And I tend to go along with that when all attributes can be 'frozen' (a `dict` attribute, like in `ParametrizedBackendFunction` is a counter-example).

Reply:
https://github.com/ComPWA/tensorwaves/pull/353#issuecomment-982770331